File: src/main/java/ure/examplegame/ExampleGame.java
package ure.examplegame;

import ure.actors.UActorCzar;
import ure.actors.UPlayer;
import ure.areas.UArea;
import ure.areas.UCartographer;
import ure.areas.UCell;
import ure.math.UColor;
import ure.render.URenderer;
import ure.render.URendererOGL;
import ure.sys.Injector;
import ure.sys.UCommander;
import ure.sys.UREGame;
import ure.sys.UTimeListener;
import ure.terrain.UTerrainCzar;
import ure.things.UThing;
import ure.things.UThingCzar;
import ure.ui.*;
import ure.ui.modals.HearModalTitleScreen;
import ure.ui.modals.UModalTitleScreen;
import ure.ui.panels.UActorPanel;
import ure.ui.panels.ULensPanel;
import ure.ui.panels.UScrollPanel;
import ure.ui.panels.UStatusPanel;

import javax.inject.Inject;

public class ExampleGame implements UREGame, HearModalTitleScreen, UTimeListener {

    static UArea area;
    static UCamera camera;
    static UPlayer player;
    static UStatusPanel statusPanel;
    static UScrollPanel scrollPanel;
    static ULensPanel lensPanel;
    static UActorPanel actorPanel;
    static URenderer renderer;

    @Inject
    UCommander commander;
    @Inject
    UTerrainCzar terrainCzar;
    @Inject
    UThingCzar thingCzar;
    @Inject
    UActorCzar actorCzar;
    @Inject
    UCartographer cartographer;

    public ExampleGame() {
        Injector.getAppComponent().inject(this);
    }

    private void makeWindow() {

        View rootView = new View();

        camera = new UCamera(renderer, 0, 0, 1200, 800);
        camera.moveTo(area, 40,20);
        rootView.addChild(camera);

        UColor borderColor = UColor.COLOR_DARKGRAY;

        statusPanel = new UStatusPanel(200, 200, 10, 10, commander.config.getTextColor(), UColor.COLOR_BLACK, borderColor);
        statusPanel.addText("name", "Kaffo",0,0);
        statusPanel.addText("race", "Owl",0,1);
        statusPanel.addText("class", "Ornithologist",0,2);
        statusPanel.addText("turn", "T 1", 0, 5);
        statusPanel.addText("time", "", 0, 6);
        statusPanel.addText("location", "?", 0, 8);
        statusPanel.addText("lens", "", 0, 20);
        statusPanel.setPosition(1200,0);
        rootView.addChild(statusPanel);

        actorPanel = new UActorPanel(200,600,10,10,commander.config.getTextColor(), UColor.COLOR_BLACK, borderColor);
        actorPanel.setPosition(1200,200);
        rootView.addChild(actorPanel);

        lensPanel = new ULensPanel(camera, 0, 0, 200, 200, 12, 12, commander.config.getTextColor(), UColor.COLOR_BLACK, borderColor);
        lensPanel.setPosition(1200,800);
        rootView.addChild(lensPanel);

        scrollPanel = new UScrollPanel(1200, 200, 12, 12, commander.config.getTextColor(), new UColor(0f,0f,0f), new UColor(0.3f,0.3f,0.3f));
        scrollPanel.addLineFade(new UColor(1.0f, 1.0f, 1.0f));
        scrollPanel.addLineFade(new UColor(0.6f, 0.6f, 0.6f));
        scrollPanel.addLineFade(new UColor(0.4f, 0.4f, 0.4f));
        scrollPanel.addLineFade(new UColor(0.3f, 0.3f, 0.3f));
        scrollPanel.setPosition(0,800);
        scrollPanel.setBounds(0,800,1200,200);
        scrollPanel.print("Welcome to UnRogueEngine!");
        scrollPanel.print("The universal java toolkit for roguelike games.");
        scrollPanel.print("Your journey begins...");
        rootView.addChild(scrollPanel);

        renderer.setRootView(rootView);



        commander.setStatusPanel(statusPanel);
        commander.setScrollPanel(scrollPanel);
        commander.registerModalCamera(camera);
        commander.config.setUiFrameGlyphs(null);
    }

    public void startUp()  {
        renderer = new URendererOGL();
        renderer.initialize();

        cartographer = new ExampleCartographer();

        commander.registerComponents(this, player, renderer, thingCzar, actorCzar, cartographer);

        makeWindow();

        commander.registerScrollPrinter(scrollPanel);
        commander.registerTimeListener(this);
        commander.addAnimator(camera);

        setupTitleScreen();

        commander.gameLoop();
    }

    public void setupTitleScreen() {
        scrollPanel.hide();
        lensPanel.hide();
        statusPanel.hide();
        actorPanel.hide();
        cartographer.setupRegions();
        area = cartographer.getTitleArea();
        if (player != null) {
            player.setCameraPinStyle(0);
            player.moveToCell(area, 50, 50);
            player = null;
        }
        camera.moveTo(area, 50, 50);
        commander.config.setVisibilityEnable(false);
        commander.showModal(new UModalTitleScreen(35, 20, this, "start", UColor.COLOR_BLACK, area));

    }

    public void hearTimeTick(UCommander cmdr) {
        if (statusPanel != null && !statusPanel.isHidden()) {
            statusPanel.setText("turn", "T " + Integer.toString(commander.getTurn()));
            statusPanel.setText("time", commander.timeString(true, " "));
            statusPanel.setText("location", commander.cartographer.describeLabel(commander.player().area().getLabel()));
        }
        if (actorPanel != null && !actorPanel.isHidden() && commander.player() != null) {
            actorPanel.updateActors((UPlayer)commander.player());
        }
    }

    public void hearModalTitleScreen(String context) {
        if (context.equals("Credits") || context.equals("Quit")) {
            commander.quitGame();
        } else {
            if (context.equals("New World")) {
                cartographer.wipeWorld();
            }
            continueGame();
        }
    }

    public void continueGame() {
        System.out.println("Creating the @Player");
        player = new UPlayer("Player", '@', UColor.COLOR_WHITE, true, new UColor(0.3f, 0.3f, 0.6f), 3, 4);
        commander.setPlayer(player);
        System.out.println("Un-hiding interface panels");
        statusPanel.unHide();
        lensPanel.unHide();
        scrollPanel.unHide();
        actorPanel.unHide();
        cartographer.startLoader();
        area = cartographer.getStartArea();
        UCell startcell = area.randomOpenCell(player);
        commander.config.setVisibilityEnable(true);
        player.attachCamera(camera, UCamera.PINSTYLE_SOFT);
        player.moveToCell(area, startcell.x, startcell.y);
        player.startActing();
        //cartographer.playerLeftArea(player, null);

        UThing item = thingCzar.getThingByName("rock");
        item.moveTo(player);
        item = thingCzar.getThingByName("trucker hat");
        item.moveTo(player);
        item = thingCzar.getThingByName("torch");
        item.moveTo(player);
        item = thingCzar.getThingByName("apple");
        item.moveTo(player);
        item = thingCzar.getThingByName("apple");
        item.moveTo(player);
        item = thingCzar.getThingByName("flashlight");
        item.moveTo(player);
        item = thingCzar.getThingByName("biscuit");
        item.moveTo(player);
        item = thingCzar.getThingByName("lantern");
        item.moveTo(player);
        item = thingCzar.getThingByName("butcher knife");
        item.moveTo(player);
        item = thingCzar.getThingByName("gold statue");
        item.moveTo(player);
        item = thingCzar.getThingByName("skull");
        item.moveTo(player);
    }
}

File: src/main/java/ure/kotlin/example/ExampleGame.kt
package ure.kotlin.example

import org.apache.commons.logging.LogFactory
import ure.actors.UActorCzar
import ure.actors.UPlayer
import ure.areas.UArea
import ure.areas.UCartographer
import ure.examplegame.ExampleCartographer
import ure.kotlin.ui.modals.HearModalTitleScreen
import ure.math.UColor
import ure.render.URenderer
import ure.sys.*
import ure.terrain.UTerrainCzar
import ure.things.UThing
import ure.things.UThingCzar
import ure.ui.UCamera
import ure.ui.modals.UModalTitleScreen
import ure.ui.panels.*
import java.io.IOException
import java.util.logging.LogManager
import javax.inject.Inject

class ExampleGame : UREgame, HearModalTitleScreen {

    var renderer: URenderer? = null
        @Inject set
    var commander: UCommander? = null
        @Inject set
    var config: UConfig? = null
        @Inject set
    var terrainCzar: UTerrainCzar? = null
        @Inject set
    var thingCzar: UThingCzar? = null
        @Inject set
    var actorCzar: UActorCzar? = null
        @Inject set

    @Inject
    lateinit var cartographer: UCartographer

    var player: UPlayer? = null

    lateinit var area: UArea
    private lateinit var camera: UCamera
    lateinit var statusPanel: StatusPanel
    lateinit var hotbarPanel: HotbarPanel
    lateinit var scrollPanel: ScrollPanel
    lateinit var lensPanel: LensPanel
    lateinit var actorPanel: ActorPanel
    lateinit var window: UREWindow

    private val log = LogFactory.getLog(ExampleGame::class.java)

    init {
        // Set up logging before doing anything else, including dependency injection.  That way we'll
        // get proper logging for @Provides methods.
        try {
            val configInputStream = ExampleGame::class.java.getResourceAsStream("/logging.properties")
            LogManager.getLogManager().readConfiguration(configInputStream)
        } catch (ioe: IOException) {
            throw RuntimeException("Can't configure logger", ioe)
        }

        Injector.appComponent.inject(this)
    }

    private fun makeWindow() {
        window = UREWindow()
        camera = UCamera(0, 0, 1200, 800).apply {
            moveTo(area, 40, 20)
            window.setCamera(this)
        }

        val borderColor = UColor.DARKGRAY

        statusPanel = StatusPanel(10, 10, config!!.textColor, null, borderColor)

        statusPanel.addText("name", " ", 0, 0)
        statusPanel.addText("race", "Owl", 0, 1)
        statusPanel.addText("class", "Ornithologist", 0, 2)
        statusPanel.addText("turn", "T 1", 0, 5)
        statusPanel.addText("time", "", 0, 6)
        statusPanel.addText("location", "?", 0, 8)
        statusPanel.addText("lens", "", 0, 20)
        statusPanel.setLayout(UPanel.XPOS_LEFT, UPanel.YPOS_BOTTOM, 8, 0.15f, 12, 10, 0f, 10)
        window.addPanel(statusPanel)

        actorPanel = ActorPanel(10, 10, config!!.textColor, null, borderColor)
        actorPanel.setLayout(UPanel.XPOS_LEFT, UPanel.YPOS_FIT, 8, 0.15f, 12, 1, 1f, 9999)
        window.addPanel(actorPanel)

        lensPanel = LensPanel(camera, 0, 0, 12, 12, config!!.textColor, null, borderColor)
        lensPanel.setLayout(UPanel.XPOS_LEFT, UPanel.YPOS_TOP, 8, 0.15f, 12, 6, 0f, 6)
        window.addPanel(lensPanel)


        scrollPanel = ScrollPanel(12, 12, config!!.textColor, null, UColor(0.3f, 0.3f, 0.3f))
        scrollPanel.setLayout(UPanel.XPOS_FIT, UPanel.YPOS_BOTTOM, 0, 1f, 9999, 2, 0.18f, 11)
        scrollPanel.addLineFade(UColor(1.0f, 1.0f, 1.0f))
        scrollPanel.addLineFade(UColor(0.8f, 0.8f, 0.8f))
        scrollPanel.addLineFade(UColor(0.6f, 0.6f, 0.6f))
        scrollPanel.addLineFade(UColor(0.5f, 0.5f, 0.5f))
        scrollPanel.addLineFade(UColor(0.4f, 0.4f, 0.4f))
        scrollPanel.addLineFade(UColor(0.3f, 0.3f, 0.3f))
        scrollPanel.print("Welcome to UnRogueEngine!")
        scrollPanel.print("The universal java toolkit for roguelike games.")
        scrollPanel.print("Your journey begins...")
        window.addPanel(scrollPanel)

        hotbarPanel = HotbarPanel(config!!.textColor, config!!.panelBgColor)
        hotbarPanel.setLayout(UPanel.XPOS_FIT, UPanel.YPOS_TOP, 1, 1f, 9999, 2, 0f, 2)
        window.addPanel(hotbarPanel)

        window.doLayout()
        renderer!!.rootView = window

        commander!!.setStatusPanel(statusPanel)
        commander!!.setScrollPanel(scrollPanel)
        commander!!.registerCamera(camera)
    }

    fun startUp() {

        cartographer = ExampleCartographer()
        makeWindow()
        commander!!.registerComponents(this, window, player, renderer, thingCzar, actorCzar, cartographer)


        commander!!.registerScrollPrinter(scrollPanel)
        commander!!.addAnimator(camera)

        setupTitleScreen()

        commander!!.gameLoop()
    }

    override fun setupTitleScreen() {
        window.hidePanels()
        area = cartographer.titleArea
        camera.moveTo(area, 50, 50)
        commander!!.config.isVisibilityEnable = false
        commander!!.showModal(UModalTitleScreen(22, 22, this, "start", area))
    }

    override fun hearModalTitleScreen(context: String, optional: String) {
        if (context == "Credits" || context == "Quit") {
            commander!!.quitGame()
        } else {
            if (context == "New World") {
                cartographer.wipeWorld()
                continueGame(optional)
            } else {
                continueGame(optional)
            }
        }
    }

    fun continueGame(playername: String) {
        area.requestCloseOut()
        player = commander!!.loadPlayer()
        if (player == null) {
            player = makeNewPlayer(playername)
            log.debug("Getting the starting area")
            cartographer.startLoader()
            area = cartographer.makeStartArea()
            val startcell = area.randomOpenCell(player)
            player!!.setSaveLocation(area, startcell.x, startcell.y)
        } else {
            log.info("Loading existing player into " + player!!.getSaveAreaLabel())
            cartographer.startLoader()
            area = cartographer.getArea(player!!.getSaveAreaLabel())
        }
        commander!!.startGame(player, area)
        player!!.attachCamera(camera, config!!.cameraPinStyle)
        window.showPanels()
    }

    fun makeNewPlayer(playername: String): UPlayer {
        log.debug("Creating a brand new @Player")
        player = UPlayer("Player", UColor(0.1f, 0.1f, 0.4f), 2, 3)
        player!!.name = playername
        player!!.id = commander!!.generateNewID(player)

        var item: UThing? = thingCzar!!.getThingByName("small stone")
        item!!.moveTo(player!!)
        item = thingCzar!!.getThingByName("trucker hat")
        item!!.moveTo(player!!)
        item = thingCzar!!.getThingByName("apple")
        item!!.moveTo(player!!)
        item = thingCzar!!.getThingByName("nylon backpack")
        item!!.moveTo(player!!)
        item = thingCzar!!.getThingByName("flashlight")
        item!!.moveTo(player!!)
        item = thingCzar!!.getThingByName("biscuit")
        item!!.moveTo(player!!)
        item = thingCzar!!.getThingByName("lantern")
        item!!.moveTo(player!!)
        item = thingCzar!!.getThingByName("butcher knife")
        item!!.moveTo(player!!)
        item = thingCzar!!.getThingByName("hiking boots")
        item!!.moveTo(player!!)

        item = thingCzar!!.getThingByName("army helmet")
        item!!.moveTo(player!!)
        item = thingCzar!!.getThingByName("confusion helmet")
        item!!.moveTo(player!!)
        item = thingCzar!!.getThingByName("aluminum bat")
        item!!.moveTo(player!!)
        item = thingCzar!!.getThingByName("leather jacket")
        item!!.moveTo(player!!)
        item = thingCzar!!.getPile("gold coins", 100)
        item!!.moveTo(player!!)
        item = thingCzar!!.getPile("gold coins", 320)
        item!!.moveTo(player!!)
        return player!!
    }
}

fun main() {
    ExampleGame().startUp()
}
